miden-lib/asm/miden/kernels/tx/note.masm

```diff
@@ -120,3 +130,157 @@ export.prepare_note
+        mem_stream hperm
+        # => [PERM, PERM, PERM, asset_data_ptr, asset_counter, num_asset_pairs, note_data_ptr]
+
+        # check if we should loop again
+        movup.13 add.1 dup movdn.14 dup.15 neq
+        # => [should_loop, PERM, PERM, PERM, asset_data_ptr, asset_counter, num_asset_pairs,
+        #     note_data_ptr]
+    end
+
+    # extract digest from hasher rate elements (h_0, ..., h_3)
+    dropw swapw dropw
+    # => [ASSETS_HASH, asset_data_ptr, asset_counter, num_asset_pairs, note_data_ptr]
+
+    # drop accessory variables from stack
+    movup.4 drop
+    movup.4 drop
+    movup.4 drop
+    # => [ASSETS_HASH, note_data_ptr]
+
+    # save vault hash to memory
+    movup.4 exec.memory::set_created_note_assets_hash
+    # => []
+end
+
+#! Computes the hash of an output note located at the specified memory address.
+#!
+#! The hash is computed as follows:
+#! - we define, recipient =
+#!       hash(hash(hash(serial_num, [0; 4]), script_hash), input_hash)
+#! - we then compute the output note hash as:
+#!       hash(recipient, assets_hash)
+#!
+#! Stack: [note_data_ptr]
+#! Output: [OUTPUT_NOTE_HASH]
+#!
+#! - note_data_ptr is a pointer to the data section of the output note.
+#! - OUTPUT_NOTE_HASH is the hash of the output note located at note_data_ptr.
+proc.compute_output_note_hash
+    # pad capacity elements of hasher
+    padw
+
+    # insert created note recipient into the first four elements of the hasher rate
+    dup.4 exec.memory::get_created_note_recipient
+
+    # populate the last four elements of the hasher rate with the created note vault hash
+    dup.8 exec.compute_output_note_assets_hash
+
+    # compute created note hash and extract digest
+    hperm dropw swapw dropw
+
+    # save created note hash to memory
+    movup.4 mem_storew
+end
+
+#! Computes a commitment to the created notes. This is computed as a sequential hash of
+#! (note_hash, note_metadata) tuples.
+#!
+#! Stack: []
+#! Output: [OUTPUT_NOTES_COMMITMENT]
+#!
+#! - OUTPUT_NOTES_COMMITMENT is the commitment to the notes created by the transaction.
+export.compute_output_notes_commitment
+    # get the number of created notes from memory
+    exec.memory::get_num_created_notes
+    # => [num_notes, ...]
+
+    # calculate the address at which we should stop looping
+    exec.memory::get_created_note_ptr
+    # => [end_ptr, ...]
+
+    # compute pointer for first address
+    push.0 exec.memory::get_created_note_ptr
+    # => [first_note_ptr, end_ptr, ...]
+
+    # prepare stack for hashing
+    padw padw padw
+    # => [PERM, PERM, PERM, first_note_ptr, end_ptr, ...]
+
+    # check if the number of created notes is greater then 0. Conditional for the while loop.
+    dup.13 dup.13 neq
+    # => [PERM, PERM, PERM, first_note_ptr, end_ptr, ...]
+
+    # loop and hash created notes
+    while.true
+        # compute created note hash (this also computes created not vault hash)
```

Comment:
should be `output note hash` and `assets hash` (instead of vault hash)?

Reply:
Yes - fixed!